### PR TITLE
fix estimated fee

### PR DESCRIPTION
**Pull request by greg-schrammel**
Fixes APP-1556
Fixes APP-1637

### CODE REVIEW

File: src/__swaps__/screens/Swap/hooks/useEstimatedGasFee.ts [hunk 1]
```diff
@@ -52,9 +53,10 @@ export function useEstimatedGasFee({
 }
 
 export function useSwapEstimatedGasFee(gasSettings: GasSettings | undefined) {
-  const { assetToSell, chainId = ChainId.mainnet, quote } = useSyncedSwapQuoteStore();
-  const { data: estimatedGasLimit, isFetching } = useSwapEstimatedGasLimit({ chainId, assetToSell, quote });
+  const { assetToSell, quote } = useSyncedSwapQuoteStore();
+  const chainId = useSwapsStore(s => s.inputAsset?.chainId || ChainId.mainnet);
 
+  const { data: estimatedGasLimit, isFetching } = useSwapEstimatedGasLimit({ chainId, assetToSell, quote });
```
Comment (christianbaroni) [inline src/__swaps__/screens/Swap/hooks/useEstimatedGasFee.ts:59]: Won't this change cause the gas limit and gas fee to be recomputed more often, and potentially behave erratically, since the `chainId` no longer changes together with the `assetToSell` and `quote`? It's also no longer guaranteed to match the `assetToSell`.

Is there some reason the `chainId` from the store is needed?

Also, can you broadly explain the changes in this PR?

Comment (greg-schrammel) [reply src/__swaps__/screens/Swap/hooks/useEstimatedGasFee.ts:59]: we start without a `quote`, so the first `estimatedGasLimit` was always a placeholder for mainnet, even if/when the input token of another chain was selected
if we don't have a quote it won't try to fetch it, just using the right chainId to have a good enough placeholder 

inside `useSwapEstimatedGasLimit` added a check to disable the query when `chainId !== assetToSell.chainId`

inside `useSwapEstimatedGasLimit` returning `{ gasLimit, chainId }` from the `queryFn` makes so we know which chain that gasLimit refers to, and we can disregard it and return a 'placeholder' if the chainId changed but the query didn't update yet

the `gasSettings` arg is updating with the inputAsset chainId, before while we didn't have a new quote it was calculating with `gasLimit` of chain `A` (because of `keepPreviousData`) but using the updated `gasSettings` of chain `B`

File: src/__swaps__/screens/Swap/hooks/useEstimatedGasFee.ts [hunk 2]
```diff
@@ -51,11 +52,15 @@ export function useEstimatedGasFee({
...
-export function useSwapEstimatedGasFee(gasSettings: GasSettings | undefined) {
-  const { assetToSell, chainId = ChainId.mainnet, quote } = useSyncedSwapQuoteStore();
-  const { data: estimatedGasLimit, isFetching } = useSwapEstimatedGasLimit({ chainId, assetToSell, quote });
+export function useSwapEstimatedGasFee(overrideGasSettings?: GasSettings) {
+  const { assetToSell, quote, chainId = ChainId.mainnet } = useSyncedSwapQuoteStore();
+  const gasSettings = useSelectedGas(chainId);
 
-  const estimatedFee = useEstimatedGasFee({ chainId, gasLimit: estimatedGasLimit, gasSettings });
+  const { data: estimatedGasLimit, isFetching } = useSwapEstimatedGasLimit({ chainId, assetToSell, quote });
+  const estimatedFee = useEstimatedGasFee({ chainId, gasLimit: estimatedGasLimit, gasSettings: overrideGasSettings || gasSettings });
```
Comment (brunobar79) [inline src/__swaps__/screens/Swap/hooks/useEstimatedGasFee.ts:60]: what happens with this hook while isFetching?

File: src/references/gasUnits.ts [hunk 3]
```diff
@@ -12,6 +12,7 @@ export const gasUnits = {
     [ChainId.zora]: '1000000',
     [ChainId.bsc]: '600000',
     [ChainId.polygon]: '600000',
+    [ChainId.avalanche]: '350000',
```
Comment (brunobar79) [inline src/references/gasUnits.ts:15]: Can you also add Degen & Blast?

File: src/__swaps__/screens/Swap/hooks/useSwapEstimatedGasLimit.ts [hunk 4]
```diff
@@ -98,8 +84,18 @@ export function useSwapEstimatedGasLimit(
...
-      placeholderData: gasUnits.basic_swap[chainId],
+      enabled: !!chainId && !!quote && !!assetToSell && assetToSell.chainId === chainId,
+      initialData,
+      initialDataUpdatedAt: 0,
       ...config,
     }
   );
+
+  // we keepPreviousData so we can return the previous gasLimit while fetching
+  // which is great when refetching for the same chainId, but we don't want to keep the previous data
```
Comment (brunobar79) [inline src/__swaps__/screens/Swap/hooks/useSwapEstimatedGasLimit.ts:95]: If I switch assets under the same chain id, I might go through a completely different route and the gas limit can be completely different, so this is not a good idea unless you take into account the whole route + chain id.
TLDR;  Only when this condition is true:
`input asset  === prev Input asset && output asset === prev output asset &&. chainId ==== prevChainId` you can return the previous value, otherwise you need to get a new one

**Review by brunobar79 — COMMENTED** [review]
Added a couple comments and there are conflicts. Once all those are fixed I'd love for @ibrahimtaveras00  to take a look to this one too

File: src/__swaps__/screens/Swap/hooks/useSwapEstimatedGasLimit.ts (hunk 4 above)
Comment (greg-schrammel) [reply src/__swaps__/screens/Swap/hooks/useSwapEstimatedGasLimit.ts:95]: its returning the previous value just while refetching the new one, and we are displaying that it's loading

File: src/__swaps__/screens/Swap/hooks/useSwapEstimatedGasLimit.ts [hunk 5]
```diff
@@ -98,8 +84,18 @@ export function useSwapEstimatedGasLimit(
       cacheTime: 60 * 1000, // 1min
       notifyOnChangeProps: ['data', 'isFetching'],
       keepPreviousData: true,
-      placeholderData: gasUnits.basic_swap[chainId],
+      enabled: !!chainId && !!quote && !!assetToSell && assetToSell.chainId === chainId,
+      initialData,
+      initialDataUpdatedAt: 0,
```
Comment (jinchung) [inline src/__swaps__/screens/Swap/hooks/useSwapEstimatedGasLimit.ts:89]: can you help me understand why you have `initialDataUpdatedAt` like this here?

Comment (greg-schrammel) [reply src/__swaps__/screens/Swap/hooks/useSwapEstimatedGasLimit.ts:89]: since I'm setting a initialData it is considered fresh so won't refetch until the stale time passes, setting it to zero makes it instantly start refetching the query

File: src/__swaps__/screens/Swap/hooks/useSwapEstimatedGasLimit.ts [hunk 6]
```diff
@@ -98,8 +84,18 @@ export function useSwapEstimatedGasLimit(
...
       ...config,
     }
   );
+
+  // we keepPreviousData so we can return the previous gasLimit while fetching
+  // which is great when refetching for the same chainId, but we don't want to keep the previous data
+  // when fetching for a different chainId
+  return {
+    data: data && data.chainId === chainId ? data.gasLimit : initialData?.gasLimit,
```
Comment (jinchung) [inline src/__swaps__/screens/Swap/hooks/useSwapEstimatedGasLimit.ts:98]: leaving a note for the future: when we'd like to upgrade to react-query v5 where we can replace this with using `placeholderData` which gets access to previous values in v5 (currently it does not, which is why we need to do this particular pattern in order to get previous data while fetching if you stay switch assets while on the same chain)




NEW

[8:16](https://rainbowhaus.slack.com/archives/D05JJTKVA0J/p1719443801054779)
which is why I presume the v5 upgrade where you can use placeholderData but also have access to the previous value is the “nicest” way to handle what we want

**Review by jinchung — APPROVED** [review]
🌮

**Review by jinchung — CHANGES_REQUESTED** [review]
RESOLVED:
Blocking PR for now - Finding weirdness in the custom gas panel - confirming if this is also happening in latest develop

Separate issue that is also happening in develop: once you update custom gas and then come back to the panel, tip resets to 0

Uploading Simulator Screen Recording - iPhone 15 Pro - 2024-06-27 at 08.43.01.mp4…

**Review by ibrahimtaveras00 — CHANGES_REQUESTED** [review]
https://rainbowhaus.slack.com/archives/C04F08AEJAZ/p1719528803523309

**Review by brunobar79 — APPROVED** [review]
LGTM
